compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmOptimizationLowering.kt

```diff
@@ -148,121 +146,59 @@ class JvmOptimizationLowering(val context: JvmBackendContext) : FileLoweringPass
-                        return rewritePrimitiveEqeqSafeCall(left, safeCallRight, expression)
+                    if (expression.symbol == context.irBuiltIns.eqeqSymbol) {
+                        if (right.type.isJvmPrimitive()) {
+                            parseSafeCall(left)?.let { return rewriteSafeCallEqeqPrimitive(it, expression) }
+                        }
+                        if (left.type.isJvmPrimitive()) {
+                            parseSafeCall(right)?.let { return rewritePrimitiveEqeqSafeCall(it, expression) }
+                        }
                     }
-
-                    return expression
                 }
 
                 return expression
             }
 
-            private fun rewriteSafeCallEqeqPrimitive(safeCall: SafeCallInfo, primitive: IrExpression, eqeqCall: IrCall): IrExpression =
-                context.createJvmIrBuilder(safeCall.scopeSymbol).run {
-                    // Fuze safe call with primitive equality to avoid boxing the primitive.
-                    // 'a?.<...> == p' becomes:
-                    //      {
-                    //          val tmp = a
-                    //          when {
-                    //              tmp == null -> false
-                    //              else -> tmp == p
-                    //          }
-                    //      }
-                    irBlock {
-                        +safeCall.tmpVal
-                        +irWhen(
-                            eqeqCall.type,
-                            listOf(
-                                irBranch(safeCall.ifNullBranch.condition, irFalse()),
-                                irElseBranch(
-                                    irCall(eqeqCall.symbol).apply {
-                                        putValueArgument(0, safeCall.ifNotNullBranch.result)
-                                        putValueArgument(1, primitive)
-                                    }
-                                )
-                            )
-                        )
-                    }
+            private fun IrBuilderWithScope.ifSafe(safeCall: SafeCallInfo, expr: IrExpression): IrExpression =
+                irBlock(origin = IrStatementOrigin.SAFE_CALL) {
+                    +safeCall.tmpVal
+                    +irIfThenElse(expr.type, safeCall.ifNullBranch.condition, irFalse(), expr)
                 }
 
-            private fun rewritePrimitiveEqeqSafeCall(primitive: IrExpression, safeCall: SafeCallInfo, eqeqCall: IrCall): IrExpression =
+            // Fuse safe call with primitive equality to avoid boxing the primitive. `a?.x == p`:
+            //     { val tmp = a; if (tmp == null) null else tmp.x } == p`
+            // is transformed to:
+            //     { val tmp = a; if (tmp == null) false else tmp.x == p }
+            // Note that the original IR implied that `p` is always evaluated, but the rewritten version
```

Comment:
This actually looks like a bug in the old BE. It's good that you've preserved the "old" behavior, though.

Reply:
Please, add a test checking that in `a?.x == expr`  `expr` is not evaluated if `a` is null.

Reply:
Done. Predictably fails on non-JVM backends - not sure whether `TARGET_BACKEND` or `IGNORE_BACKEND` more appropriate, this is a language design issue after all.